### PR TITLE
doc: add initial doc for shard shrinking

### PR DESCRIPTION
**Pull request by smakalias**
## Summary of the changes / Why this improves CrateDB

Document the new shard shrinking API 

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: blackbox/docs/general/ddl/alter-table.rst [hunk 1]
```diff
@@ -32,6 +32,35 @@ In order to set a parameter to its default value use ``reset``::
...
+-------------------------
+
+It is possible to change the number of primary shards of a table
+
+..    todo: remove comments once feature is implemented
+..    cr> alter table my_table set (number_of_shards = 2);
+..    ALTER OK, -1 rows affected (... sec)
+
+.. Note::
+    Currently we only support table shrinking, meaning that
```
Comment (seut) [inline blackbox/docs/general/ddl/alter-table.rst]: I'd remove the `Currently`

File: blackbox/docs/general/ddl/partitioned-tables.rst [hunk 2]
```diff
@@ -372,9 +372,19 @@ both existing and future partitions of the table.
...
     ALTER OK, -1 rows affected (... sec)
 
-Unlike regular tables, it is possible to change the number of shards of a
-partitioned table. However, the shard configuration will only be applied on
-**future** partitions.
+
+Changing Number of Shards
+-------------------------
+
+it is possible at any time to change the number of shards
```
Comment (seut) [inline blackbox/docs/general/ddl/partitioned-tables.rst]: Capital letter

File: blackbox/docs/general/ddl/partitioned-tables.rst [hunk 3]
```diff
@@ -372,9 +372,19 @@ both existing and future partitions of the table.
...
+of a partitioned table. Currently, we support two options.
```
Comment (seut) [inline blackbox/docs/general/ddl/partitioned-tables.rst]: I'd change this to sections *Altering whole partitioned table*  and *Altering a single partition*

File: blackbox/docs/general/ddl/alter-table.rst [hunk 4]
```diff
@@ -32,6 +32,35 @@ In order to set a parameter to its default value use ``reset``::
...
+.. _alter_change_number_of_shard:
+
+Changing Number of Shards
+-------------------------
+
+It is possible to change the number of primary shards of a table
+
+..    todo: remove comments once feature is implemented
+..    cr> alter table my_table set (number_of_shards = 2);
+..    ALTER OK, -1 rows affected (... sec)
```
Comment (seut) [inline blackbox/docs/general/ddl/alter-table.rst]: We want this will to fail until implementation is done. (enable it, remove comments)

File: blackbox/docs/sql/statements/create-table.rst [hunk 5]
```diff
@@ -224,7 +224,7 @@ Several restrictions apply to columns that can be used here:
 .. _with_clause:
 
 ``WITH``
---------
+========
```
Comment (seut) [inline blackbox/docs/sql/statements/create-table.rst]: Afaik the changes in this file are unrelated to this PR, I'd prefer having them in a dedicated PR or at least commit.

File: blackbox/docs/general/ddl/alter-table.rst [hunk 6]
```diff
@@ -32,6 +32,35 @@ In order to set a parameter to its default value use ``reset``::
     cr> alter table my_table reset (number_of_replicas);
     ALTER OK, -1 rows affected (... sec)
 
+.. _alter_change_number_of_shard:
+
+Changing Number of Shards
```
Comment (andreidan) [inline blackbox/docs/general/ddl/alter-table.rst]: "Changing The Number ..."

File: blackbox/docs/general/ddl/alter-table.rst [hunk 7]
```diff
@@ -32,6 +32,35 @@ In order to set a parameter to its default value use ``reset``::
...
+..    ALTER OK, -1 rows affected (... sec)
+
+.. Note::
+    Currently we only support table shrinking, meaning that
+    the requested number of shards must be less than the
+    current number of the primary shards. Moreover, the requested number
+    of primary shards must be a factor of the current number of primary shards.
+    For example, a table with 8 primary shards can be shrunk
+    into 4, 2 or 1 primary shards.
+    If the number of shards of a table is a prime number,
```
Comment (andreidan) [inline blackbox/docs/general/ddl/alter-table.rst]: I think this should a warning on its own as it has bigger implications than a "note" would imply

File: blackbox/docs/general/ddl/partitioned-tables.rst [hunk 8]
```diff
@@ -423,6 +431,34 @@ partitioned table. However, the shard configuration will only be applied on
...
+..    | partition_ident          | values                 | number_of_shards |
+..    +--------------------------+------------------------+------------------+
+..    | 04732cpp6osj2d9i60o30c1g | {"day": 1396915200000} |                5 |
+..    +--------------------------+------------------------+------------------+
+..    SELECT 1 row in set (... sec)
+
+.. NOTE::
+
+   This option operates on a partition basis, thus a partition ident needs to
+   be specified. In addition, same prerequisites and restrictions as with normal
```
Comment (andreidan) [inline blackbox/docs/general/ddl/partitioned-tables.rst]: "the same prerequisites"

File: blackbox/docs/general/ddl/partitioned-tables.rst (hunk 3 above)
Comment (smakalias) [reply blackbox/docs/general/ddl/partitioned-tables.rst]: I chose the specific wording in order to stress out the difference (existing vs future).
with your proposal this is not stressed enough (at least not from the title) and the "whole partitioned table" is ab bit misleading as well, right? It only applies for the future partitions.

My reasoning was to make that crystal clear from the start.


Let me know your thoughts

File: blackbox/docs/sql/statements/create-table.rst (hunk 5 above)
Comment (smakalias) [reply blackbox/docs/sql/statements/create-table.rst]: valid point

File: blackbox/docs/general/ddl/alter-table.rst (hunk 1 above)
Comment (smakalias) [reply blackbox/docs/general/ddl/alter-table.rst]: done

File: blackbox/docs/general/ddl/alter-table.rst (hunk 7 above)
Comment (smakalias) [reply blackbox/docs/general/ddl/alter-table.rst]: done

File: blackbox/docs/general/ddl/alter-table.rst [hunk 9]
```diff
@@ -32,6 +32,38 @@ In order to set a parameter to its default value use ``reset``::
...
+-----------------------------
+
+It is possible to change the number of primary shards of a table
+::
+
+    cr> alter table my_table set (number_of_shards = 2);
+    ALTER OK, -1 rows affected (... sec)
+
+.. Warning::
+    For changing the number of primary shards,
```
Comment (andreidan) [inline blackbox/docs/general/ddl/alter-table.rst]: I'd phrase this as "In order to change the number of primary shards..."

File: blackbox/docs/general/ddl/partitioned-tables.rst [hunk 10]
```diff
@@ -372,9 +372,13 @@ both existing and future partitions of the table.
...
-Unlike regular tables, it is possible to change the number of shards of a
-partitioned table. However, the shard configuration will only be applied on
-**future** partitions.
+
+Changing the Number of Shards
+-----------------------------
+
+It is possible at any time to change the number of shards
+of a partitioned table. However, the shard configuration
+will only be applied on -**future** partitions.
```
Comment (andreidan) [inline blackbox/docs/general/ddl/partitioned-tables.rst]: "be applied to" instead of "on"

also, there's an extra "-" character before **future**. I'd also maybe replace **future** with **new** partitions? (we use "new" below as well and I think it makes more sense)

File: blackbox/docs/general/ddl/sharding.rst [hunk 11]
```diff
@@ -46,8 +46,10 @@ is applied, (see :ref:`ref_clustered_clause`).
 
 .. NOTE::
 
-   The number of shards can only be set on table creation, it cannot be changed
-   later on.
+   The number of shards can be changed later on but it involves having the table
+   in a ``READ-ONLY`` state until the operation is completed
+   (see :ref:`alter_change_number_of_shard`).
+   So, it is better to tune this carefully in advance.
```
Comment (andreidan) [inline blackbox/docs/general/ddl/sharding.rst]: "So, it is better to tune this carefully in advance." - maybe it's a personal preference, but I find this a bit condescending. I wouldn't keep it in the documentation.

File: blackbox/docs/general/ddl/partitioned-tables.rst (hunk 10 above)
Comment (smakalias) [reply blackbox/docs/general/ddl/partitioned-tables.rst]: +1 for "new" partitions

File: blackbox/docs/general/ddl/alter-table.rst [hunk 12]
```diff
@@ -32,6 +32,38 @@ In order to set a parameter to its default value use ``reset``::
...
+-----------------------------
+
+It is possible to change the number of primary shards of a table.
+::
+
+    cr> alter table my_table set (number_of_shards = 2);
+    ALTER OK, -1 rows affected (... sec)
+
+.. Warning::
+    In order to change the number of primary shards,
```
Comment (seut) [inline blackbox/docs/general/ddl/alter-table.rst]: pls add a newline

File: blackbox/docs/general/ddl/alter-table.rst [hunk 13]
```diff
@@ -32,6 +32,38 @@ In order to set a parameter to its default value use ``reset``::
...
+    the table must be explicitly set to `READ-ONLY` first
+    (see :ref:`table-settings-blocks.read_only` parameter),
+    otherwise the operation will fail.
+    In addition, this parameter will have to be explicitly restored
+    to it's original value after the operation.
+
+.. Note::
+    We only support table shrinking, meaning that
```
Comment (seut) [inline blackbox/docs/general/ddl/alter-table.rst]: here also
